### ISSUE TITLE
`skeema lint` deletes comments within "CREATE TABLE" statements

### ISSUE THREAD

**Issue by RJPercival** [skeema/skeema#72]
I have "CREATE TABLE" statements that look like the following:
```sql
CREATE TABLE `Foo` (
  `Id` bigint(20) NOT NULL,
  -- Bar is blah-blah-blah.
  `Bar` varbinary(255) NOT NULL,
  PRIMARY KEY (`Id`)
) ENGINE=InnoDB CHARSET=utf8mb4;
```

After running `skeema lint`, the "-- Bar is blah-blah-blah" comment is deleted. Is it possible to preserve comments?

**Comment by evanelias** [comment]
Currently Skeema's reformatting logic is just using the canonical format from `SHOW CREATE TABLE`, which does not include mid-statement SQL comments, since MySQL does not store them. Having Skeema maintain mid-statement comments is possible in theory, but it would require a full SQL parser, which has been [intentionally avoided](https://github.com/skeema/skeema/blob/master/doc/faq.md#no-reliance-on-sql-parsing) so far.

Two current work-arounds, although they're not ideal:
* SQL comments before or after the full `CREATE TABLE` statement are preserved
* Comments specified using a COMMENT clause, e.g. `` `Bar` varbinary(255) NOT NULL COMMENT 'blah-blah-blah'``, are preserved

If you just want to lint schemas without reformatting anything, two other work-arounds will be available in the near future:
* An option will be added to `skeema lint` to tell it not to reformat files
* The [pull request auto-linter](https://www.skeema.io/blog/2019/01/19/github-api-beta/) won't reformat things (it doesn't request permissions to write to the repo).

**Comment by RJPercival** [comment]
Thank you for your comprehensive response. I tried out using a COMMENT clause and that works, though it does result in indices that have gained comments being deleted and recreated by Skeema, which in our case could take quite a while (some of our tables have 100,000,000+ rows). I haven't come across any benefits we get from storing the comments in that way, so I'll probably just move the comments out of the `CREATE TABLE` statement instead.

It would be great to have an option to disable reformatting. Could it still write the reformatted file elsewhere, e.g. to stdout? That'd help with applying some of the reformatting manually. I guess what I'm actually looking for is a flag to control whether it writes to stdout or to the original file (e.g. the `-w` flag for `gofmt`).

**Comment by evanelias** [comment]
Agreed re: `COMMENT` clauses; I generally advise against their use for this reason.

Regarding an option to disable reformatting in lint, my current thinking is along these lines:
* Add a separate `skeema format` command, which only reformats without including any linter annotations
* `skeema lint` would still reformat files by default, but would have a `--skip-format` option to disable this behavior
* Both commands would get a new `--display-format-diff` option, which behaves similarly to `gofmt -d`.

These changes are currently planned to coincide with some other linter-related tweaks, such as having `skeema diff` and `skeema push` perform linting by default, skipping any dir with lint errors. Currently targeting this for v1.3.0 in one to two months, but open to PRs for individual pieces.

**Comment by evanelias** [comment]
Forgot to close this old issue a while back. [`skeema lint --skip-format`](https://www.skeema.io/docs/options/#format) and the separate [`skeema format` command](https://www.skeema.io/docs/commands/format) were implemented in v1.3.0 from Sept 2019.
